src/widget/file_navigator/directory_view.rs

```diff
@@ -196,48 +197,52 @@ impl<'a> Widget for DirectoryView<'a> {
             });
 
             let show_hidden = self.show_hidden;
-            let entries: Vec<_> = match std::fs::read_dir(directory).ok() {
+            let mut entries: Vec<_> = match std::fs::read_dir(directory).ok() {
                 Some(entries) => {
                     entries.filter_map(|e| e.ok())
-                        .filter(|f| check_hidden(show_hidden, &f.path()))
-                        .collect()
+                        .filter_map(|f| {
+                            let path = f.path();
+                            if check_hidden(show_hidden, &path) {
+                                match types {
+                                    super::Types::All => Some(path),
+                                    super::Types::WithExtension(valid_exts) => {
+                                        // We're only after files.
+                                        if path.is_dir() {
+                                            Some(path)
+                                        } else {
+                                            // Check for valid extensions.
+                                            let ext = path.extension()
+                                                .and_then(|ext| ext.to_str())
+                                                .map(|s| std::ascii::AsciiExt::to_ascii_lowercase(s))
+                                                .unwrap_or_else(String::new);
+                                            if valid_exts.iter().any(|&valid_ext| &ext == valid_ext) {
+                                                Some(path)
+                                            } else {
+                                                None
+                                            }
+                                        }
+                                    },
+                                }
+                            } else {
+                                None
+                            }
+                        }).collect()
                 }
```

Comment:
The original code did things in a slightly strange way. First it created a Vec with all the entries. Then it created an iterator that went through the Vec once selecting all the files. Then it created another iterator that went through the Vec again selecting all the files that matched. And then it joined the two iterators. In the end it was doing several Vec traversals. If I had kept that code, I would have to do a collect() on that final iterator to sort again. That's what I did original and the code was just getting too complex. If it's the rightward drift you're worried it's easy enough to just move the contents of the filter_map into it's own function together with check_hidden.